pkg/ingester/index.go

```diff
@@ -77,16 +77,17 @@ func (shard *indexShard) lookup(matchers []*labels.Matcher) []model.Fingerprint
 		if !ok {
 			return nil
 		}
-		var toIntersect []model.Fingerprint
+		var toIntersect model.Fingerprints
 		if matcher.Type == labels.MatchEqual {
 			fps := values[model.LabelValue(matcher.Value)]
-			toIntersect = merge(toIntersect, fps)
+			toIntersect = append(toIntersect, fps...)
 		} else {
 			for value, fps := range values {
 				if matcher.Matches(string(value)) {
-					toIntersect = merge(toIntersect, fps)
+					toIntersect = append(toIntersect, fps...)
 				}
 			}
+			sort.Sort(toIntersect)
```

Comment:
It seems like the ideal algorithm would be a k-way merge?

Reply:
Have you got one?

Reply:
Not one at my fingertips. I think I am good with this as is, we can always reduce the time complexity later if we want/need to.

Just brainstorming - since all we care about is actually the intersection, we could keep `result` as a map which would mean we don't even need to append in many cases. Then sort once before returning. I'd be happy to try that later though.

Reply:
I think it's very case-by-case.
* If the query is something like `{namespace="default"}`, then we'll go through the loop once, never intersect.
* For `{namespace="default",container_name="nginx"}` we'll intersect two pre-existing sorted slices - way more efficient than constructing a `map` then deleting from it.
* The `append()` and `Sort()` will be used on a query like `{namespace="default",container_name!=""}`, and whether a map would be faster depends on the specific data you are holding.